src/QsCompiler/Transformations/CallGraph/CallGraphWalkerBase.cs

```diff
@@ -0,0 +1,87 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.Quantum.QsCompiler.DataTypes;
+using Microsoft.Quantum.QsCompiler.DependencyAnalysis;
+using Microsoft.Quantum.QsCompiler.SyntaxTree;
+using Microsoft.Quantum.QsCompiler.Transformations.Core;
+
+namespace Microsoft.Quantum.QsCompiler.Transformations.CallGraphWalker
+{
+    using Range = DataTypes.Range;
+    using TypeParameterResolutions = ImmutableDictionary<Tuple<QsQualifiedName, NonNullable<string>>, ResolvedType>;
+
+    internal static partial class BuildCallGraph
+    {
+        private static class CallGraphWalkerBase<TGraph, TNode, TEdge>
+            where TGraph : CallGraphBuilder<TNode, TEdge>
+            where TNode : CallGraphNodeBase
+            where TEdge : CallGraphEdgeBase
+        {
+            public abstract class TransformationState
+            {
+                internal TNode? CurrentNode;
+                internal readonly TGraph Graph;
+
+                // The type parameter resolutions of the current expression.
+                internal IEnumerable<TypeParameterResolutions> ExprTypeParamResolutions = new List<TypeParameterResolutions>();
```

Comment:
Since you are prepending, a List here seems like a bit of a suboptimal choice. You could go with a Stack instead.

Reply:
done